keps/sig-scheduling/2926-job-mutable-scheduling-directives/README.md

```diff
@@ -0,0 +1,836 @@
+
+Adding the ability to mutate a job's node affinity and tolerations gives a queue controller
+the ability to influence placement while at the same time offloading actual pod-to-node
+assignment to the existing kube-scheduler.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Allow mutating node affinity and tolerations of suspended jobs.
+
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Implement a queue controller.
+- Allow mutating node affinity or tolerations of running jobs.
+- Allow mutating node affinity or tolerations of pods.
+- Allow mutating node affinity or tolerations of apps other than jobs.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+The proposal is to relax validation of suspended jobs' update to the pod template's
+node affinity and tolerations.
+
+This has no impact on the job-controller. The job controller has no dependency
+on the node affinity constraints or tolerations expressed in a job's pod template.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to build a controller that implements job queueing and influences when and where a 
+job should run. Users create v1.Job objects, and to control when the job can run, I have 
+a webhook that forces the jobs to be created in a suspended state, and the controller unsuspends 
+them when capacity becomes available.
+
+At the time of creating the job, we may not know on which part of the cluster (.e.g., a zone or 
+a VM type) a job should run. To control where a job should run, the controller can update 
+node affinity or tolerations of the Job's pod template; by doing so, the queue controller is 
+able to influence the scheduler and autoscaler decisions.
+
+<!--
+#### Story 2
+
+
+### Notes/Constraints/Caveats (Optional)
+-->
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```

Comment:
Let's also add a risk that was discussed in k/k issue about the races of mutating a job that was suspended but not all of its pods were deleted.

Reply:
I commented on the issue, if we add the condition active=0, then i don't think there is one.  I didn't add this condition to the KEP yet, wanted to wait to see if there are other potential race scenarios.

Reply:
There are races - Aldo pointed out one, there are other too.

Reply:
added.